### DISCUSSION TITLE
[라이프로그] activity 데이터에서 Dementia 의 표본 갯수가 부족함에서 발생하는 문제

### DISCUSSION THREAD

**Discussion by yonghyeun** [yonghyeun/Dementia_Analysis#5] [General]
```python
REPEATE = 10

fig,ax = plt.subplots(ncols = 5, nrows = 2, 
                        figsize = (15,10))
ax = ax.flatten()

X = df.drop('target',axis = 1)
Y = df['target']
    
avg_score = 0    

for i in range(REPEATE):
    

    x_train,x_test,y_train,y_test = train_test_split(X,Y,random_state = 42 + i, stratify = Y)
    
    y_train , y_test = one_hot_encoder(y_train),one_hot_encoder(y_test)
    
    model = create_dnn(**params)
    model.compile(**optimizers) 
    
    history = model.fit(x_train,y_train,
                validation_data = (x_test,y_test),
                epochs = 9999,
                callbacks = [early_stopping],
                verbose = 0)
    
    score = model.evaluate(x_test,y_test,verbose = 0)[-1]
    
    avg_score += score / REPEATE
    
    ax[i].plot(history.history['accuracy'], c = 'orange',label = 'train_accuracy')
    ax[i].plot(history.history['val_accuracy'], c = 'green',label = 'valid_accuracy')
    
    ax[i].set_title(f'{i + 1} FOLD SET \n ACCURACY : {round(score,3)}', fontsize = 10)
    
    
fig.suptitle(f'10 FOLD TEST \n AVG ACCURACY : {round(avg_score,2)}')


color = ['orange','green']
label = ['train_accuracy','valid_accuracy']
legend_color = {'train_accuracy':'orange','valid_accuracy':'green'}


legend_elements = [plt.Line2D([0], [0], marker='o', color=color, label=label) 
                    for label, color in legend_color.items()]

fig.legend(handles=legend_elements,
            bbox_to_anchor = (0.7,1),
            frameon = False)
```
![8c1af465-a7a7-4eb3-8aaf-7d9ac5229a43](https://user-images.githubusercontent.com/123540354/234896662-7b25c56d-1688-4213-b072-f2fdc509e05f.png)

해당 코드에서 교차검증 시 테스트 데이터가 바뀜에도 불구하고 대부분의 accuracy 값이 비슷한 값에 수렴하는 문제 발생

실험해보니 stratify 를 지정을 해주느냐 안해주느냐에 따라서 accuracy 값의 변화가 생기는데 지금 생각하고 있는 예상으로는 현재 모델이 Dementia 의 갯수가 적으니 (y_test 에서 34개 중 총 2개) 전부 0으로 밀어버리고 치매 초기군과 비 치매 환자들의 값들만 예측 하고 있는 것으로 보임

익일 시간 내어 각 모델들의 예측값이 어떻게 되는지 확인해봐야 할 거 같음

**Comment by yonghyeun** [comment]
![image](https://user-images.githubusercontent.com/123540354/235080629-9d59ed56-c592-46bf-aca4-371266315081.png)


생각하다보니 비치매군, 초기 치매군, 치매군 환자간의 데이터 특성이 독립 변수들로 명확하게 나타나지 않아 생기는 문제가 아닐까? 라는 생각이 들기도 한다.

데이터간 특성의 차이가 없으니, 모델은 가장 많은 갯수를 가지고 있는 비 치매군으로 전부 밀어버리고 있는 거 같다. 

feature engineering 을 통해서 정확도를 나중에 올려봐야겠다.

**Comment by 2mook2** [comment]
threshold 를 조정해 보는 것은 어떨까요~?

**Reply by yonghyeun** [reply]
참고해서 해보겠습니다 ^^
